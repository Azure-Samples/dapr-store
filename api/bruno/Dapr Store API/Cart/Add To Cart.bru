meta {
  name: Add To Cart
  type: http
  seq: 1
}

put {
  url: {{endpoint}}/cart/method/setProduct/{{userId}}/prd001/6
  body: none
  auth: inherit
}

assert {
  res.status: eq 200
  res.body: isJson
  res.body.products: isNotEmpty
}
