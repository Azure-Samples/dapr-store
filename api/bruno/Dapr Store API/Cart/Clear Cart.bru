meta {
  name: Clear Cart
  type: http
  seq: 4
}

put {
  url: {{endpoint}}/cart/method/clear/{{userId}}
  body: none
  auth: inherit
}

assert {
  res.status: eq 200
  res.body: isJson
  res.body.products: isEmpty
}
