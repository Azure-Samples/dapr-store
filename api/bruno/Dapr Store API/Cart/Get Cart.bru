meta {
  name: Get Cart
  type: http
  seq: 2
}

get {
  url: {{endpoint}}/cart/method/get/{{userId}}
  body: none
  auth: inherit
}

assert {
  res.status: eq 200
  res.body: isJson
  res.body.products: isNotEmpty
}
