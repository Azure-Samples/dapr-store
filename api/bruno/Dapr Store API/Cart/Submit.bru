meta {
  name: Submit
  type: http
  seq: 3
}

post {
  url: {{endpoint}}/cart/method/submit
  body: text
  auth: inherit
}

body:text {
  "{{userId}}"
}

vars:post-response {
  orderId: res.body.id
}

assert {
  res.status: eq 200
  res.body: isJson
  res.body.amount: isNumber
  res.body.lineItems: isArray
  res.body.status: eq new
}
