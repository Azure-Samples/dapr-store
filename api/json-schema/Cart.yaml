$schema: https://json-schema.org/draft/2020-12/schema
$id: Cart.yaml
type: object
properties:
  products:
    $ref: "#/$defs/RecordInt32"
  forUserId:
    type: string
required:
  - products
  - forUserId
$defs:
  RecordInt32:
    type: object
    properties: {}
    unevaluatedProperties:
      type: integer
      minimum: -2147483648
      maximum: 2147483647
