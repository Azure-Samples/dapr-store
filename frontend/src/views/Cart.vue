<template>
  <div>
    <h1><fa icon="shopping-cart" /> &nbsp; Shopping Cart</h1><br>
    <b-table striped hover :items="user.cart" :fields="fields" />
    <b-button :disabled="user.cart.length == 0" variant="primary" size="lg">
      <fa icon="shopping-basket" /> &nbsp; CHECKOUT
    </b-button>
  </div>
</template>

<script>
import { userProfile } from '../main'

export default {
  name: 'Cart',

  data() {
    return {
      fields: ['name', 'cost', 'description'],
      user: userProfile
    }
  },
}
</script>

<style>

</style>