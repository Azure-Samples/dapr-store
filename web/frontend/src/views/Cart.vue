<!--
// ----------------------------------------------------------------------------
// Copyright (c) Ben Coleman, 2020
// Licensed under the MIT License.
//
// Dapr Store frontend - Cart contents view
// ----------------------------------------------------------------------------
-->

<template>
  <div>
    <h1>
      <fa icon="shopping-cart" /> &nbsp; Shopping Cart
    </h1>
    <br>
    <error-box :error="error" />

    <div v-if="!cart && !error" class="text-center">
      <b-spinner variant="success" style="width: 5rem; height: 5rem;" />
    </div>

    <b-alert v-if="newOrder" show variant="success">
      <h4>Order accepted! ðŸ˜„</h4>
      <b-link :to="`order/`+newOrder.id">
        <div>
          Your order for <b>Â£{{ newOrder.amount }}</b> has been accepted, the order ID is: <b>{{ newOrder.id }}</b><br>
          Check your account for progress on your order(s)
        </div>
      </b-link>
    </b-alert>

    <b-alert v-if="cart && Object.keys(cart.products).length==0" show variant="dark">
      Your cart is empty! <br><br><b-link href="/catalog">
        Go shopping!
      </b-link>
    </b-alert>

    <div v-if="cart">
      <h2>Cart total: Â£{{ total.toFixed(2) }} </h2>
      <b-card v-for="product of cartProducts" :key="product.id" class="m-3 p-1" header-bg-variant="primary" header-text-variant="white">
        <template v-slot:header>
          <span>{{ product.name }}</span>
        </template>
        <h2>Count:</h2> <input :value="cart.products[product.id]" type="text" readonly>

        <b-button class="ml-5 mr-3" :disabled="!cart || cart.products.length == 0" variant="warning" size="lg" @click="modifyProductAmmount(product.id, -1)">
          <fa icon="minus-circle" />
        </b-button>
        <b-button :disabled="!cart || cart.products.length == 0" variant="success" size="lg" @click="modifyProductAmmount(product.id, 1)">
          <fa icon="plus-circle" />
        </b-button>
        <img :src="product.image" class="thumb">
      </b-card>
    </div>

    <b-button :disabled="!cart || cartProducts.length == 0" variant="primary" size="lg" @click="submitOrder">
      <fa icon="shopping-basket" /> &nbsp; CHECKOUT
    </b-button>
    &nbsp;
    <b-button :disabled="!cart || cartProducts.length == 0" variant="warning" size="lg" class="float-right" @click="clearCart">
      <fa icon="trash-alt" /> &nbsp; EMPTY CART
    </b-button>
  </div>
</template>

<script>
import ErrorBox from '../components/ErrorBox'
import api from '../services/api'
import auth from '../services/auth'

export default {
  name: 'Cart',

  components: {
    'error-box': ErrorBox
  },

  data() {
    return {
      error: null,
      newOrder: null,
      cart: null,
      cartProducts: [],
      user: null
    }
  },

  computed: {
    total() {
      let tot = 0
      for (let product of this.cartProducts) {
        let count = this.cart.products[product.id]
        tot += (count * product.cost)
      }

      return tot
    }
  },

  async mounted() {
    try {
      this.user = auth.user()
      if (!this.user) { return }

      let resp = await api.cartGet(this.user.userName)
      if (resp) {
        this.cart = resp
        this.cartProducts = []
        for (let productId in this.cart.products) {
          // Do this async helps speed it up when running locally, due to Dapr bug
          api.productGet(productId)
            .then((resp) => {
              this.cartProducts.push(resp)
            })
        }
      }
    } catch (err) {
      this.error = err
    }
  },

  methods: {
    async submitOrder() {
      try {
        this.newOrder = await api.cartSubmit(this.user.userName)
        this.cart = await api.cartClear(this.user.userName)
        this.cartProducts = []
      } catch (err) {
        this.error = err
      }
    },

    async clearCart() {
      try {
        let resp = await api.cartClear(this.user.userName)
        this.cart = resp
        this.cartProducts = []
      } catch (err) {
        this.error = err
      }
    },

    async modifyProductAmmount(productId, amount) {
      try {
        this.cart = await api.cartAddAmount(this.user.userName, productId, amount)

        // Fiddly nonsense to remove from cartProducts if removed from products.cart
        // Check if productId is removed from cart object, then recreate cartProducts array
        if (!Object.prototype.hasOwnProperty.call(this.cart.products, productId)) {
          this.cartProducts = this.cartProducts.filter((p) => p.id != productId)
        }
      } catch (err) {
        this.error = err
      }
    }
  }
}
</script>

<style scoped>
  input[type=text] {
    border: none;
    width: 4rem;
    text-align: center;
  }
  .thumb {
    width: 80px;
    float: right;
  }
</style>