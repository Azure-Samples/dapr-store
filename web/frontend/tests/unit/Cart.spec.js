import { createLocalVue, mount } from '@vue/test-utils'
import flushPromises from 'flush-promises'

import BootstrapVue from 'bootstrap-vue'
const localVue = createLocalVue()
localVue.use(BootstrapVue)

import VueRouter from 'vue-router'
localVue.use(VueRouter)
const router = new VueRouter()

import Cart from '@/views/Cart.vue'

jest.mock('@/services/api')
jest.mock('@/services/auth')


describe('Cart.vue', () => {
  it('renders cart contents', async () => {
    const wrapper = mount(Cart, {
      localVue,
      router
    })

    await flushPromises()
    expect(wrapper).toMatchSnapshot()
  })
})