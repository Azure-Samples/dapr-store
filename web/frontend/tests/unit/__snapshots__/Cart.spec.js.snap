// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart.vue renders cart contents 1`] = `
<div>
  <h1>
    <fa icon="shopping-cart"></fa> &nbsp; Shopping Cart
  </h1> <br>
  <!---->
  <!---->
  <!---->
  <!---->
  <div>
    <h2>Cart total: £91.10 </h2>
    <div class="card m-3 p-1">
      <!---->
      <div class="card-header bg-primary text-white"><span>Top&nbsp;Hat&nbsp;(6″)</span></div>
      <div class="card-body">
        <!---->
        <!---->
        <h2>Count:</h2> <input type="text" readonly="readonly"> <button type="button" class="btn ml-5 mr-3 btn-warning btn-lg">
          <fa icon="minus-circle"></fa>
        </button> <button type="button" class="btn btn-success btn-lg">
          <fa icon="plus-circle"></fa>
        </button> <img src="/img/catalog/1.jpg" class="thumb">
      </div>
      <!---->
      <!---->
    </div>
    <div class="card m-3 p-1">
      <!---->
      <div class="card-header bg-primary text-white"><span>Paisley Cravat Ascot Tie</span></div>
      <div class="card-body">
        <!---->
        <!---->
        <h2>Count:</h2> <input type="text" readonly="readonly"> <button type="button" class="btn ml-5 mr-3 btn-warning btn-lg">
          <fa icon="minus-circle"></fa>
        </button> <button type="button" class="btn btn-success btn-lg">
          <fa icon="plus-circle"></fa>
        </button> <img src="/img/catalog/3.jpg" class="thumb">
      </div>
      <!---->
      <!---->
    </div>
  </div> <button type="button" class="btn btn-primary btn-lg">
    <fa icon="shopping-basket"></fa> &nbsp; CHECKOUT
  </button>
  &nbsp;
  <button type="button" class="btn float-right btn-warning btn-lg">
    <fa icon="trash-alt"></fa> &nbsp; EMPTY CART
  </button>
</div>
`;
